dimensiondata/src/main/java/org/jclouds/dimensiondata/cloudcontrol/config/DimensionDataCloudControlComputeServiceContextModule.java

```diff
@@ -0,0 +1,254 @@
+            pollPeriod.pollInitialPeriod, pollPeriod.pollMaxPeriod);
+   }
+
+   @Provides
+   @Named(SERVER_DELETED_PREDICATE)
+   @VisibleForTesting
+   public Predicate<String> provideServerDeletedPredicate(final DimensionDataCloudControlApi api,
+         final ComputeServiceConstants.Timeouts timeouts, final ComputeServiceConstants.PollPeriod pollPeriod) {
+      return retry(new ServerState(api.getServerApi(), State.DELETED), timeouts.imageAvailable,
+            pollPeriod.pollInitialPeriod, pollPeriod.pollMaxPeriod);
+   }
+
+   @Provides
+   @Named(SERVER_NORMAL_PREDICATE)
+   protected Predicate<String> provideServerNormalPredicate(final DimensionDataCloudControlApi api,
+         final ComputeServiceConstants.Timeouts timeouts, final ComputeServiceConstants.PollPeriod pollPeriod) {
+      return retry(new ServerState(api.getServerApi(), State.NORMAL), timeouts.imageAvailable,
+            pollPeriod.pollInitialPeriod, pollPeriod.pollMaxPeriod);
+   }
+
+   @Provides
+   @Named(VM_TOOLS_RUNNING_PREDICATE)
+   protected Predicate<String> provideVMToolsRunningPredicate(final DimensionDataCloudControlApi api,
+         final ComputeServiceConstants.Timeouts timeouts, final ComputeServiceConstants.PollPeriod pollPeriod) {
+      return retry(new VMToolsRunningStatus(api.getServerApi()), timeouts.imageAvailable, pollPeriod.pollInitialPeriod,
+            pollPeriod.pollMaxPeriod);
+   }
+
+   private class VlanState implements Predicate<String>
+
+   {
+      private final State state;
+      private final NetworkApi networkApi;
+
+      private VlanState(final NetworkApi networkApi, final State state) {
+         this.networkApi = networkApi;
+         this.state = state;
+      }
+
+      @Override
+      public boolean apply(final String vlanId) {
+         checkNotNull(vlanId, "vlanId");
+         logger.trace("looking for state on vlan %s", vlanId);
+         final Vlan vlan = networkApi.getVlan(vlanId);
+         final boolean isDeleted = (vlan == null) && (state == State.DELETED);
+         return isDeleted || ((vlan != null) && vlan.state() == state);
+      }
+   }
+
+   private class NetworkDomainState implements Predicate<String> {
+
+      private final State state;
+      private final NetworkApi networkApi;
+
+      private NetworkDomainState(final NetworkApi networkApi, final State state) {
+         this.networkApi = networkApi;
+         this.state = state;
+      }
+
+      @Override
+      public boolean apply(final String networkDomainId) {
+         checkNotNull(networkDomainId, "networkDomainId");
+         logger.trace("looking for state on network domain %s", networkDomainId);
+         final NetworkDomain networkDomain = networkApi.getNetworkDomain(networkDomainId);
+         final boolean isDeleted = networkDomain == null && state == State.DELETED;
+         return isDeleted || (networkDomain != null && networkDomain.state() == state);
+      }
+   }
```

Comment:
If this predicate pattern is meant to be used in more classes, consider extracting it to a base class.You could create an interface `WithStatus` (or whatever) that provides the `state()` method, and that would allow you to write a generic predicate that checks the status with just an abstract method to get the `WithStatus` object to be checked. Just an idea to avoid duplicating code if the pattern is going to be the same for all objects with state.

Reply:
Makes sense, for now we should not require more of these Predicates. But I might take this on during my next set of refactoring.